xarray/plot/plot.py

```diff
@@ -919,24 +919,31 @@ def imshow(x, y, z, ax, **kwargs):
     The pixels are centered on the coordinates. For example, if the coordinate
     value is 3.2, then the pixels for those coordinates will be centered on 3.2.
     """
+    if any(np.issubdtype(v.dtype, str) for v in [x, y]):
+        raise TypeError("ax.imshow does not support categorical data.")
 
     if x.ndim != 1 or y.ndim != 1:
         raise ValueError(
             "imshow requires 1D coordinates, try using pcolormesh or contour(f)"
         )
 
-    # Centering the pixels- Assumes uniform spacing
-    try:
-        xstep = (x[1] - x[0]) / 2.0
-    except IndexError:
-        # Arbitrary default value, similar to matplotlib behaviour
-        xstep = 0.1
-    try:
-        ystep = (y[1] - y[0]) / 2.0
-    except IndexError:
-        ystep = 0.1
-    left, right = x[0] - xstep, x[-1] + xstep
-    bottom, top = y[-1] + ystep, y[0] - ystep
+    def _maybe_center_pixels(x):
+        """Center the pixels on the coordinates."""
+        if np.issubdtype(x.dtype, str):
+            # Categorical data cannot be centered:
+            return None, None
+
+        try:
+            # Center the pixels assuming uniform spacing:
+            xstep = 0.5 * (x[1] - x[0])
+        except IndexError:
+            # Arbitrary default value, similar to matplotlib behaviour:
+            xstep = 0.1
+        return x[0] - xstep, x[-1] + xstep
+
+    # Try to center the pixels:
+    left, right = _maybe_center_pixels(x)
```

Comment:
This makes sure that xarray isn't the one throwing errors before getting to ax.imshow. 
But it's true it's not really necessary as ax.imshow doesn't seem to support categoricals.

But maybe we prefer matplotlib raising the error?

Reply:
> np.issubdtype(v.dtype, str) check hadn't been there.

but it is there :) It's redundant, is it not? Shall we just remove the str check from `_maybe_center_pixels`?

Reply:
The other alternative is to remove the other check and let it crash within matplotlib. But I'm not sure they even intend to support categoricals for these functions.

Reply:
I think i like that. If matplotlib starts supporting things, then it will just work automatically in xarray. I also liked your refactoring to `_maybe_center_pixels`

Reply:
Ok, then I just need to figure out how to catch the errors in the tests.